### PR TITLE
add marker trait RmiiClkPin.

### PR DESCRIPTION
**Pull request by elwerene**
implement trait with config for gpio0, gpio16, gpio17

### CODE REVIEW

File: src/gpio.rs [hunk 1]
```diff
@@ -123,6 +123,12 @@ pub trait TouchPin: Pin {
     fn touch_channel(&self) -> touch_pad_t;
 }
 
+/// A marker trait designating a pin which is capable of
```
Comment (ivmarkov) [inline src/gpio.rs:126]: This marker trait is ESP32 specific, so it has to be enabled only when compiling for the original ESP32 chip with `#[cfg(esp32)]. Also everywhere below.

File: src/gpio.rs [hunk 2]
```diff
@@ -123,6 +123,12 @@ pub trait TouchPin: Pin {
...
+/// providing/receiving a rmii clk signal
+pub trait RmiiClkPin: Pin {
+    fn clock_config() -> eth_mac_clock_config_t__bindgen_ty_2;
```
Comment (ivmarkov) [inline src/gpio.rs:129]: `eth_mac_clock_config_t__bindgen_ty_2` looks... yeah. Do we need *just* this part of the union, or can we switch to `eth_mac_clock_config_t`?

If we really can't how about `pub type <something more meaningful> =  eth_mac_clock_config_t__bindgen_ty_2`;

File: src/gpio.rs (hunk 1 above)
Comment (ivmarkov) [reply src/gpio.rs:126]: In fact, it also needs to be protected not to compile when the `riscv-ulp-hal` feature is enabled. So:
```
#[cfg(esp32)]
#[cfg(not(feature = "riscv-ulp-hal"))]
```

File: src/gpio.rs [hunk 3]
```diff
@@ -1076,6 +1082,40 @@ mod chip {
     pin!(Gpio38:38, Input, RTC:2, ADC1:2, NODAC:0, NOTOUCH:0);
     pin!(Gpio39:39, Input, RTC:3, ADC1:3, NODAC:0, NOTOUCH:0);
 
+    impl<MODE> RmiiClkPin for Gpio0<MODE>
```
Comment (ivmarkov) [inline src/gpio.rs:1085]: All of these protected with `#[cfg(esp32)]` as per above. You can also probably move them to the esp32-specific `mod chip` and then re-export them from there with `pub use...`

File: src/gpio.rs [hunk 4]
```diff
@@ -1076,6 +1082,40 @@ mod chip {
...
+    where
+        MODE: Send,
```
Comment (ivmarkov) [inline src/gpio.rs:1087]: Nit: I don't think you need the `Send` constraint here?

**Review by ivmarkov — CHANGES_REQUESTED** [review]

File: src/gpio.rs (hunk 2 above)
Comment (elwerene) [reply src/gpio.rs:129]: I didn't want to copy all the contents of `eth_mac_clock_config_t` that's why I returned only the parts that change.

File: src/gpio.rs (hunk 1 above)
Comment (elwerene) [reply src/gpio.rs]: moved to esp32 chip and added the feature guard

File: src/gpio.rs (hunk 4 above)
Comment (elwerene) [reply src/gpio.rs:1089]: I copied it, so no :)

Comment (elwerene) [reply src/gpio.rs:1089]: actually, it is needed for pin to be send. I just tried without

File: src/gpio.rs (hunk 3 above)
Comment (elwerene) [reply src/gpio.rs:1087]: they where already in the esp32 `mod chip`. I added the guard for risc..

File: src/gpio.rs (hunk 2 above)
Comment (ivmarkov) [reply src/gpio.rs]: `eth_mac_clock_config_t` is defined like this:
```
#[doc = " @brief Ethernet MAC Clock Configuration"]
#[doc = ""]
#[repr(C)]
#[derive(Copy, Clone)]
pub union eth_mac_clock_config_t {
    #[doc = "< EMAC MII Clock Configuration"]
    pub mii: eth_mac_clock_config_t__bindgen_ty_1,
    #[doc = "< EMAC RMII Clock Configuration"]
    pub rmii: eth_mac_clock_config_t__bindgen_ty_2,
}
#[repr(C)]
#[derive(Debug, Default, Copy, Clone)]
pub struct eth_mac_clock_config_t__bindgen_ty_1 {}
#[repr(C)]
#[derive(Debug, Copy, Clone)]
pub struct eth_mac_clock_config_t__bindgen_ty_2 {
    #[doc = "< RMII Clock Mode Configuration"]
    pub clock_mode: emac_rmii_clock_mode_t,
    #[doc = "< RMII Clock GPIO Configuration"]
    pub clock_gpio: emac_rmii_clock_gpio_t,
}
```

... so you *are* already copying "all the contents". Let's switch to `eth_mac_clock_config_t`. And let's somehow make all of this compatible with ESP-IDF 4.3. E.g. an option would be to have this trait only for ESP-IDF >= 4.4 (<=> !(ESP.IDF = 4.3)) and then in `esp-idf-svc` on V4.3 hard-code gpio0 somehow, as it used to be.

Comment (ivmarkov) [reply src/gpio.rs]: ... or if that's easier, keep the Clk trait (I like it!), but on V4.3 leave it out as a marker trait only, and then on V4.3 make gpio0 the only one implementing it.

Comment (elwerene) [reply src/gpio.rs]: you are right, I was refering to `eth_mac_config_t`

Comment (elwerene) [reply src/gpio.rs]: I changed it so it is only implemented for gpipo16/17 on esp idf != 4.3

**Pull request closed without merging**